File: 0_1.baekjoon/brute-force/java/src/BOJ21943.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class BOJ21943 {
    private static int N, P, Q;
    private static int[] arr;
    private static int[] ps;
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        N = Integer.parseInt(br.readLine());
        arr = new int[N];
        ps = new int[N];

        StringTokenizer st = new StringTokenizer(br.readLine());

        for (int i = 0; i < N; i++) {
            arr[i] = Integer.parseInt(st.nextToken());
        }

        st = new StringTokenizer(br.readLine());

        P = Integer.parseInt(st.nextToken());
        Q = Integer.parseInt(st.nextToken());

        System.out.println(dfs(0));
    }

    private static int dfs(int idx) {
        if (idx == N) {
            return getResult();
        }

        int result = 0;

        for (int i = 0; i <= Q; i++) {
            ps[i] += arr[idx];
            result = Math.max(result, dfs(idx + 1));
            ps[i] -= arr[idx];
        }

        return result;
    }

    private static int getResult() {
        int result = 1;

        for (int i = 0; i <= Q; i++) {
            result *= ps[i];
        }

        return result;
    }
}

File: 0_1.baekjoon/brute-force/kotlin/src/BOJ21943.kt
import kotlin.math.max

var n = 0
var p = 0
var q = 0
lateinit var arr: IntArray
lateinit var ps: IntArray
fun main() = with(System.`in`.bufferedReader()) {
    n = readLine().toInt()
    arr = readLine().split(" ").map { it.toInt() }.toIntArray()
    readLine().split(" ")
        .also { result ->
            p = result[0].toInt()
            q = result[1].toInt()
        }


    ps = IntArray(q + 1) { 0 }

    print(dfs(0))
}

private fun dfs(idx: Int): Int {
    if (idx == n) {
        return getResult()
    }

    var result = 0
    for (i in 0 .. q) {
        ps[i] += arr[idx]
        result = max(result, dfs(idx + 1))
        ps[i] -= arr[idx]
    }

    return result
}

private fun getResult(): Int {
    var result = 1
    for (i in 0 .. q) {
        result *= ps[i]
    }

    return result
}

